### PR TITLE
i#3203: Add a basic last-level cache (LLC) misses analyzer

### PR DESCRIPTION
**Pull request by kharbutli**
The analyzer first identifies load instructions that suffer from significant last-level cache (LLC) misses. The analyzer then analyzes the cache line address streams for each of those load instructions in search for patterns that can be used for SW prefetching. The analyzer here can only find simple strides. Analyses for more complex patterns can be added in the future.
A subsequent PR will connect the changes here to the front-end via new runtime options.

Issue #3203

### CODE REVIEW

File: clients/drcachesim/simulator/cache_misses_analyzer.h [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+#include "cache_simulator.h"
+#include "cache_stats.h"
+#include "../common/memref.h"
+
+// Represents the SW prefetching recommendation passed to the compiler.
+struct prefetching_recommendation_t {
+    uint64_t pc;      // Load instruction's address.
+    int64_t stride;   // Prefetching stride/delta distance.
+    std::string locality;  // Prefetching locality: one of "nta" or "t0".
```
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_misses_analyzer.h]: Should this be an enum, or should those 2 strings be static const string values?

File: clients/drcachesim/simulator/cache_misses_analyzer.h [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    // The function returns a nonzero stride value if it finds one that
+    // satisies the confidence threshold and returns 0 otherwise.
+    int64_t
+    check_for_constant_stride(const std::vector<addr_t>& cache_misses) const;
+
+    // A hash map storing the data cache line addresses accessed by load
+    // instructions that miss in the LLC.
+    // Key is the PC of the load instruction.
+    // Value is a vector of data memory cache line addresses.
+    std::map<addr_t, std::vector<addr_t>> pc_cache_misses;
```
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_misses_analyzer.h]: This can be `std::unordered_map`

File: clients/drcachesim/simulator/cache_misses_analyzer.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    return recommendations;
+}
+
+int64_t
+cache_misses_stats_t::check_for_constant_stride(
+    const std::vector<addr_t>& cache_misses) const
+{
+    std::map<int64_t, int64_t> stride_counts;
```
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_misses_analyzer.cpp]: This can be `std::unordered_map`

File: clients/drcachesim/simulator/cache_misses_analyzer.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                                       misses_frac_threshold,
+                                       confidence_threshold);
+  llcaches["LL"]->set_stats(ll_stats_);
+}
+
+std::vector<prefetching_recommendation_t*>
+cache_misses_analyzer_t::generate_recommendations()
+{
+    return ll_stats_->generate_recommendations();
+}
```
Comment (derekbruening) [inline clients/drcachesim/simulator/cache_misses_analyzer.cpp]: I think you forgot to include changes to the cmake code?  These files are never built.

Also, I assume there will be runtime option changes to allow the user to select this: will they be in a separate CL?  That seems fine since the test here will be built and testing the code until then.

File: clients/drcachesim/simulator/cache_misses_analyzer.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "cache_misses_analyzer.h"
```
[the remark this answers is not in the document]
Comment (kharbutli) [reply clients/drcachesim/simulator/cache_misses_analyzer.cpp]: Done.

File: clients/drcachesim/simulator/cache_misses_analyzer.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+
+cache_misses_stats_t::cache_misses_stats_t(bool warmup_enabled,
+                                           uint64_t line_size,
+                                           uint64_t misses_count_threshold,
```
[the remark this answers is not in the document]
Comment (kharbutli) [reply clients/drcachesim/simulator/cache_misses_analyzer.cpp]: Done.

File: clients/drcachesim/simulator/cache_misses_analyzer.h [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+
+    // Total number of LLC misses added to the hash map above.
+    int64_t total_misses = 0;
+};
+
+class cache_misses_analyzer_t : public cache_simulator_t {
+public:
+    // Constructor:
+    // - cache_simulator_knobs_t encapsulates the cache simulator params.
```
[the remark this answers is not in the document]
Comment (kharbutli) [reply clients/drcachesim/simulator/cache_misses_analyzer.h]: Done.

**Comment by derekbruening** [comment]
> Please mention that this will be connected to the front-end via new runtime options in a future change.

Just to clarify, I meant to say to please mention this in the commit message.
